pmd-core/src/main/java/net/sourceforge/pmd/RuleBuilder.java

```diff
@@ -0,0 +1,246 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.lang3.StringUtils;
+import org.w3c.dom.Element;
+
+import net.sourceforge.pmd.lang.Language;
+import net.sourceforge.pmd.lang.LanguageRegistry;
+import net.sourceforge.pmd.lang.LanguageVersion;
+import net.sourceforge.pmd.properties.PropertyDescriptor;
+
+
+/**
+ * Builds a rule, validating its parameters throughout. The builder can define property descriptors, but not override
+ * them. For that, use {@link RuleFactory#decorateRule(Rule, Element)}.
+ *
+ * @author Clément Fournier
+ * @since 6.0.0
+ */
+class RuleBuilder {
+
+    private List<PropertyDescriptor<?>> definedProperties = new ArrayList<>();
+    private String name;
+    private String clazz;
+    private Language language;
+    private String minimumVersion;
+    private String maximumVersion;
+    private String since;
+    private String message;
+    private String rulesetName;
+    private String externalInfoUrl;
+    private String description;
+    private List<String> examples = new ArrayList<>(1);
+    private RulePriority priority;
+    private boolean isDeprecated;
+    private boolean isUsesDfa;
+    private boolean isUsesMultifile;
+    private boolean isUsesTyperesolution;
+
+
+    RuleBuilder(String name, String clazz, String language) {
```

Comment:
if this is intended to be package-private, document it appropriately https://pmd.github.io/pmd-5.8.1/pmd-java/rules/java/comments.html#CommentDefaultAccessModifier